mlir/lib/Target/IRDLToCpp/IRDLToCpp.cpp

```diff
@@ -0,0 +1,559 @@
+//===- IRDLToCpp.cpp - Converts IRDL definitions to C++ -------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Target/IRDLToCpp/IRDLToCpp.h"
+#include "mlir/Dialect/IRDL/IR/IRDL.h"
+#include "mlir/Support/LLVM.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "TemplatingUtils.h"
+
+using namespace mlir;
+
+constexpr char headerTemplateText[] =
+#include "Templates/Header.txt"
+    ;
+
+constexpr char declarationMacroFlag[] = "GEN_DIALECT_DECL_HEADER";
+constexpr char definitionMacroFlag[] = "GEN_DIALECT_DEF";
+
+namespace {
+
+struct DialectStrings {
+  StringRef dialectName;
+  StringRef dialectCppName;
+  StringRef dialectCppShortName;
+  StringRef dialectBaseTypeName;
+
+  StringRef namespaceOpen;
+  StringRef namespaceClose;
+  StringRef namespacePath;
+};
+
+struct TypeStrings {
+  StringRef typeName;
+  std::string typeCppName;
+};
+
+struct OpStrings {
+  StringRef opName;
+  std::string opCppName;
+  llvm::SmallVector<std::string> opResultNames;
+  llvm::SmallVector<std::string> opOperandNames;
+};
+
+static std::string joinNameList(llvm::ArrayRef<std::string> names) {
+  std::string nameArray;
+  llvm::raw_string_ostream nameArrayStream(nameArray);
+  nameArrayStream << "{\"" << llvm::join(names, "\", \"") << "\"}";
+
+  return nameArray;
+}
+
+static std::string snakeToCamel(llvm::StringRef in) {
+  std::string output{};
+  output.reserve(in.size());
+  for (size_t i = 0; i < in.size(); ++i) {
+    if (in[i] == '_' && i + 1 < in.size()) {
+      output += toupper(in[i + 1]);
+      ++i;
+    } else {
+      output += in[i];
+    }
+  }
+  return output;
+}
+
+static std::string snakeToPascal(llvm::StringRef in) {
+  std::string output{static_cast<std::string::value_type>(toupper(in.front()))};
+  output.reserve(in.size());
+  for (size_t i = 1; i < in.size(); ++i) {
+    if (in[i] == '_' && i + 1 < in.size()) {
+      output += toupper(in[i + 1]);
+      ++i;
+    } else {
+      output += in[i];
+    }
+  }
+  return output;
+}
```

Comment:
It could also be nice to enforce all names are valid camel case in the IRDL verifiers too.

```suggestion
static std::string snakeToCamel(llvm::StringRef in, bool capitalize = false) {
  std::string output;
  output.reserve(in.size());

  for (char c : in) {
    if (c == '_') {
      capitalize = true;
    } else {
      output += capitalize ? toupper(c) : c;
      capitalize = false;
    }
  }
  return output;
}

static std::string snakeToPascal(llvm::StringRef in) {
    return snakeToCamel(in, /*capitalize=*/true);
}
```